scylla-cdc/src/cdc_reader.rs

```diff
@@ -0,0 +1,213 @@
+                            err = Some(error);
+                            CDCReaderWorker::set_upper_timestamp(&self.readers, chrono::Duration::min_value()).await;
+                        },
+                        _ => {}
+                    }
+                }
+                Some(generation) = generation_receiver.recv(), if next_generation.is_none() => {
+                    next_generation = Some(generation.clone());
+                    had_first_generation = true;
+                    CDCReaderWorker::set_upper_timestamp(&self.readers, generation.timestamp).await;
+                }
+                Ok(_) = self.end_timestamp_receiver.changed() => {
+                    let timestamp = *self.end_timestamp_receiver.borrow_and_update();
+                    self.end_timestamp = timestamp;
+                    CDCReaderWorker::set_upper_timestamp(&self.readers, timestamp).await;
+                }
+            }
+
+            if stream_reader_tasks.is_empty() {
+                if err.is_some() {
+                    return Err(err.unwrap());
+                }
+
+                if next_generation.is_some() {
+                    let generation = next_generation.unwrap();
+                    next_generation = None;
+
+                    if generation.timestamp > self.end_timestamp {
+                        return Ok(());
+                    }
+
+                    self.readers = fetcher
+                        .fetch_stream_ids(&generation)
+                        .await?
+                        .iter()
+                        .flatten()
+                        .map(|stream_id| {
+                            Arc::new(StreamReader::new(
+                                &self.session,
+                                vec![stream_id.clone()],
+                                self.start_timestamp,
+                                self.window_size,
+                                self.safety_interval,
+                                self.sleep_interval,
+                            ))
+                        })
+                        .collect();
+
+                    CDCReaderWorker::set_upper_timestamp(&self.readers, generation.timestamp).await;
+
+                    stream_reader_tasks = self
+                        .readers
+                        .iter()
+                        .map(|reader| {
+                            let reader = Arc::clone(reader);
+                            let keyspace = self.keyspace.clone();
+                            let table_name = self.table_name.clone();
+                            let consumer = self.consumer_factory.new_consumer();
+                            tokio::spawn(async move {
+                                reader
+                                    .fetch_cdc_with_consumer(keyspace, table_name, consumer)
+                                    .await
+                            })
+                        })
+                        .collect();
```

Comment:
I didn't manage to create a new consumer asynchronously, so I removed the `async` trait from `ConsumerFactory`. An alternative is to use [block_on](https://docs.rs/futures/0.2.0/futures/executor/fn.block_on.html), but I think that in this way we show the user that it doesn't matter if the consumers can be created asynchronously.

Reply:
Also, the spawning forced me to require consumers to implement `Send`. Just like with `ConsumerFactory` (see my other comment).

Reply:
Consumers can be `Send`, it's practically required when you work with `tokio` and want an object alive across `.await` points because `tokio`'s scheduler is allowed to move tasks between worker threads.

I'd recommend putting the `Sync` and `Send` requirements in the traits' definitions in order to simplify things:

```rust
#[async_trait]
pub trait Consumer: Send {
    /* ... */
}

#[async_trait]
pub trait ConsumerFactory: Sync {
    /* ... */
}
```